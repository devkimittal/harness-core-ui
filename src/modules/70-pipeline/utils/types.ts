/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { StringNGVariable, NumberNGVariable, SecretNGVariable, PipelineInfoConfig } from 'services/cd-ng'
import type { NodeRunInfo, EntityGitDetails, EntityValidityDetails, InputSetResponse } from 'services/pipeline-ng'

export type AllNGVariables = StringNGVariable | NumberNGVariable | SecretNGVariable

export interface ExecutionPageQueryParams {
  view?: 'log' | 'graph'
  stage?: string
  step?: string
  retryStep?: string
}

export interface ExpressionBlock {
  expression?: string
  expressionValue?: string
}

export interface ConditionalExecutionNodeRunInfo extends NodeRunInfo {
  expressions?: ExpressionBlock[]
}

export type StatusType = 'SUCCESS' | 'FAILURE' | 'ERROR'

export type CreateUpdateInputSetsReturnType = Promise<{
  status?: StatusType
  nextCallback: () => void
}>

export interface InputSetType {
  name: string
  tags?: { [key: string]: string }
  identifier: string
  description: string
  orgIdentifier: string
  projectIdentifier: string
  pipeline: PipelineInfoConfig
  gitDetails: EntityGitDetails
  entityValidityDetails: EntityValidityDetails
}

export interface InputSetDTO extends Omit<InputSetResponse, 'identifier' | 'pipeline'> {
  pipeline?: PipelineInfoConfig
  identifier?: string
  repo?: string
  branch?: string
}

export interface SaveInputSetDTO {
  inputSet: InputSetDTO
}
