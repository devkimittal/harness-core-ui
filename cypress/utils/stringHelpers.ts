export function getIdentifierFromName(str: string): string {
  return str
    .trim()
    .replace(/[^0-9a-zA-Z_$ ]/g, '') // remove special chars except _ and $
    .replace(/ +/g, '_') // replace spaces with _
}
