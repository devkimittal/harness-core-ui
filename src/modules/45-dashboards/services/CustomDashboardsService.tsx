/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { SetStateAction } from 'react'
import { useGet, useMutate, GetDataError } from 'restful-react'

export interface TagsResponseDataResource {
  tags?: string
}

export interface TagsResponseData {
  resource?: TagsResponseDataResource
}

export interface TagsResponse {
  data?: TagsResponseData
  loading?: boolean
}

export const folderIdOrBlank = (folderId: string): string => (folderId === 'shared' ? '' : folderId)

export const useGetTags = (accountId: string, folderId: string): TagsResponse => {
  const { data: response, loading: isLoading } = useGet({
    path: `gateway/dashboard/v1/tags`,
    queryParams: {
      accountId: accountId,
      folderId: folderIdOrBlank(folderId)
    }
  })
  return { data: response, loading: isLoading }
}

export interface FolderDetailResponseData {
  resource?: string
}

export interface FolderDetailResponse {
  data?: FolderDetailResponseData
}

export const useGetFolderDetail = (accountId: string, folderId: string): FolderDetailResponse => {
  const { data: folderDetail } = useGet({
    path: 'gateway/dashboard/folderDetail',
    queryParams: { accountId: accountId, folderId: folderIdOrBlank(folderId) }
  })
  return { data: folderDetail }
}

export interface DashboardDetailResponseData {
  title: string
}

export interface DashboardDetailResponse {
  data: DashboardDetailResponseData
}

export const useGetDashboardDetail = (accountId: string, viewId: string): DashboardDetailResponse => {
  const { data: dashboardDetail } = useGet({
    // Inferred from RestfulProvider in index.js
    path: `gateway/dashboard/${viewId}/detail`,
    queryParams: { accountId: accountId }
  })
  return { data: dashboardDetail }
}

export interface CreateSignedUrlResponseResource {
  resource: SetStateAction<string>
}

export interface CreateSignedUrlResponse {
  mutate: (props: any) => Promise<CreateSignedUrlResponseResource>
  loading: boolean
  error?: GetDataError<any> | null
}

export const useMutateCreateSignedUrl = (
  accountId: string,
  viewId: string,
  host: string,
  query: string
): CreateSignedUrlResponse => {
  const { mutate, loading, error } = useMutate({
    verb: 'POST',
    path: 'gateway/dashboard/v1/signedUrl',
    queryParams: {
      accountId: accountId,
      src: `/embed/dashboards-next/${viewId}?embed_domain=${host}&${query}`
    }
  })
  return { mutate, loading, error }
}

export interface CloneDashboardResponseResource {
  id: string
  folder_id: string
}

export interface CloneDashboardResponse {
  mutate: (props: { dashboardId: string }) => Promise<CloneDashboardResponseResource>
  loading: boolean
}

export const useCloneDashboard = (accountId: string): CloneDashboardResponse => {
  const { mutate, loading } = useMutate({
    verb: 'POST',
    path: 'gateway/dashboard/clone',
    queryParams: {
      accountId: accountId
    }
  })
  return { mutate, loading }
}

export interface DeleteDashboardResponseResource {
  id: string
  folder_id: string
}

export interface DeleteDashboardResponse {
  mutate: (props: { dashboardId: string }) => Promise<DeleteDashboardResponseResource>
  loading: boolean
}

export const useDeleteDashboard = (accountId: string): DeleteDashboardResponse => {
  const { mutate, loading } = useMutate({
    verb: 'DELETE',
    path: 'gateway/dashboard/remove',
    queryParams: {
      accountId: accountId
    }
  })
  return { mutate, loading }
}

export interface MoveDashboardResponseData {
  resource: string
}

export interface MoveDashboardResponse {
  mutate: (props: { dashboardId: string; folderId: string }) => Promise<MoveDashboardResponseData>
  loading: boolean
}

export const useMoveDashboard = (accountId: string): MoveDashboardResponse => {
  const { mutate, loading } = useMutate({
    verb: 'POST',
    path: 'gateway/dashboard/move',
    queryParams: {
      accountId: accountId
    }
  })
  return { mutate, loading }
}

export interface UpdateDashboardResponseData {
  resource: string
}

export interface UpdateDashboardResponse {
  mutate: (props: UpdateDashboard) => Promise<UpdateDashboardResponseData>
  loading: boolean
}

export interface UpdateDashboard {
  dashboardId: string
  title: string
  description: string
  folderId: string
}

export const useUpdateDashboard = (accountId: string): UpdateDashboardResponse => {
  const { mutate, loading } = useMutate({
    verb: 'PATCH',
    path: 'gateway/dashboard',
    queryParams: {
      accountId: accountId
    }
  })
  return { mutate, loading }
}
