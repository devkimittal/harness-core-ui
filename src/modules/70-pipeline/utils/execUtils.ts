import type { PipelineGraphState } from '@pipeline/components/AbstractNode/types'
import type { ExecutionPipelineNode } from '@pipeline/components/ExecutionStageDiagram/ExecutionPipelineModel'
import type { ExecutionGraph, ExecutionNode } from 'services/pipeline-ng'
import {
  StepGroupRollbackIdentifier,
  NodeType,
  RollbackContainerCss,
  getIconDataBasedOnType,
  LITE_ENGINE_TASK,
  getExecutionPipelineNodeType,
  processLiteEngineTask
} from './executionUtils'
import type { ExecutionStatus } from './statusHelpers'

export const processNodeDataV1 = (
  children: string[],
  nodeMap: ExecutionGraph['nodeMap'],
  nodeAdjacencyListMap: ExecutionGraph['nodeAdjacencyListMap'],
  rootNodes: PipelineGraphState[]
): any[] => {
  const items: any[] = []
  children?.forEach(item => {
    const nodeData = nodeMap?.[item]
    const isRollback = nodeData?.name?.endsWith(StepGroupRollbackIdentifier) ?? false
    if (nodeData?.stepType === NodeType.FORK) {
      const [parentNode, ...childNodes] = nodeAdjacencyListMap?.[item].children as any[]

      const iconData = getIconDataBasedOnType(parentNode)
      items.push({
        id: item as string,
        identifier: item as string,
        name: parentNode?.name as string,
        icon: iconData.icon,
        type: 'StepGroup',
        nodeType: getExecutionPipelineNodeType(parentNode?.stepType),
        data: {
          name: parentNode.name || /* istanbul ignore next */ '',
          identifier: item,
          data: parentNode,
          containerCss: {
            ...(isRollback ? RollbackContainerCss : {})
          },
          status: parentNode.status as ExecutionStatus,
          isOpen: true,
          skipCondition: parentNode.skipInfo?.evaluatedCondition ? parentNode.skipInfo.skipCondition : undefined,
          when: parentNode.nodeRunInfo,
          ...iconData
        },
        children: processNodeDataV1(
          childNodes || /* istanbul ignore next */ [],
          nodeMap,
          nodeAdjacencyListMap,
          rootNodes
        )
      })
    } else if (
      nodeData?.stepType === NodeType.STEP_GROUP ||
      nodeData?.stepType === NodeType.NG_SECTION ||
      (nodeData && isRollback)
    ) {
      const iconData = getIconDataBasedOnType(nodeData)
      items.push({
        id: item as string,
        identifier: item as string,
        name: nodeData?.name as string,
        icon: iconData.icon,
        type: 'StepGroup',
        nodeType: getExecutionPipelineNodeType(nodeData?.stepType),
        data: {
          name: nodeData.name || /* istanbul ignore next */ '',
          identifier: item,
          data: nodeData,
          containerCss: {
            ...(isRollback ? RollbackContainerCss : {})
          },
          status: nodeData.status as ExecutionStatus,
          isOpen: true,
          skipCondition: nodeData.skipInfo?.evaluatedCondition ? nodeData.skipInfo.skipCondition : undefined,
          when: nodeData.nodeRunInfo,
          ...getIconDataBasedOnType(nodeData)
        },
        children: processNodeDataV1(
          nodeAdjacencyListMap?.[item].children || /* istanbul ignore next */ [],
          nodeMap,
          nodeAdjacencyListMap,
          rootNodes
        )
      } as PipelineGraphState)
    } else {
      if (nodeData?.stepType === LITE_ENGINE_TASK) {
        const parentNodeId =
          Object.entries(nodeAdjacencyListMap || {}).find(([_, val]) => {
            return (val?.children?.indexOf(nodeData.uuid!) ?? -1) >= 0
          })?.[0] || ''
        processLiteEngineTask(nodeData, rootNodes, nodeMap?.[parentNodeId])
      } else {
        const iconData = getIconDataBasedOnType(nodeData)
        items.push({
          id: item as string,
          identifier: item as string,
          name: nodeData?.name as string,
          icon: iconData.icon,
          type: nodeData?.stepType,
          nodeType: getExecutionPipelineNodeType(nodeData?.stepType),
          data: {
            name: nodeData?.name || /* istanbul ignore next */ '',
            ...iconData,
            identifier: item,
            skipCondition: nodeData?.skipInfo?.evaluatedCondition ? nodeData?.skipInfo.skipCondition : undefined,
            when: nodeData?.nodeRunInfo,
            status: nodeData?.status as ExecutionStatus,
            type: getExecutionPipelineNodeType(nodeData?.stepType),
            data: nodeData
          }
        } as PipelineGraphState)
      }
    }
    const nextIds = nodeAdjacencyListMap?.[item].nextIds || /* istanbul ignore next */ []
    nextIds.forEach(id => {
      const nodeDataNext = nodeMap?.[id]
      const isRollbackNext = nodeDataNext?.name?.endsWith(StepGroupRollbackIdentifier) ?? false
      if (nodeDataNext?.stepType === NodeType.FORK) {
        // items.push({
        //   parallel: processNodeDataV1(
        //     nodeAdjacencyListMap?.[id].children || /* istanbul ignore next */ [],
        //     nodeMap,
        //     nodeAdjacencyListMap,
        //     rootNodes
        //   )
        // })

        const [parentNode, ...childNodes] = nodeAdjacencyListMap?.[id].children as any[]

        const iconData = getIconDataBasedOnType(parentNode)
        items.push({
          id: item as string,
          identifier: item as string,
          name: parentNode?.name as string,
          icon: iconData.icon,
          type: 'StepGroup',
          nodeType: getExecutionPipelineNodeType(parentNode?.stepType),
          data: {
            name: parentNode.name || /* istanbul ignore next */ '',
            identifier: item,
            data: parentNode,
            containerCss: {
              ...(isRollback ? RollbackContainerCss : {})
            },
            status: parentNode.status as ExecutionStatus,
            isOpen: true,
            skipCondition: parentNode.skipInfo?.evaluatedCondition ? parentNode.skipInfo.skipCondition : undefined,
            when: parentNode.nodeRunInfo,
            ...iconData
          },
          children: processNodeDataV1(
            childNodes || /* istanbul ignore next */ [],
            nodeMap,
            nodeAdjacencyListMap,
            rootNodes
          )
        })
      } else if (nodeDataNext?.stepType === NodeType.STEP_GROUP || (isRollbackNext && nodeDataNext)) {
        // type: 'StepGroup',
        // nodeType: 'StepGroup',

        const iconData = getIconDataBasedOnType(nodeDataNext)
        items.push({
          id: item as string,
          identifier: item as string,
          name: nodeData?.name as string,
          icon: iconData.icon,
          type: 'StepGroup',
          nodeType: 'StepGroup',
          data: {
            name: nodeDataNext.name || /* istanbul ignore next */ '',
            identifier: id,
            data: nodeDataNext,
            containerCss: {
              ...(isRollbackNext ? RollbackContainerCss : {})
            },
            skipCondition: nodeDataNext.skipInfo?.evaluatedCondition ? nodeDataNext.skipInfo.skipCondition : undefined,
            when: nodeDataNext.nodeRunInfo,
            status: nodeDataNext.status as ExecutionStatus,
            isOpen: true,
            ...iconData
          },
          children: processNodeDataV1(
            nodeAdjacencyListMap?.[id].children || /* istanbul ignore next */ [],
            nodeMap,
            nodeAdjacencyListMap,
            rootNodes
          )
        })
      } else {
        const iconData = getIconDataBasedOnType(nodeDataNext)
        items.push({
          id: id as string,
          identifier: id as string,
          name: nodeDataNext?.name as string,
          icon: iconData.icon,
          type: nodeDataNext?.stepType,
          nodeType: getExecutionPipelineNodeType(nodeDataNext?.stepType),
          data: {
            name: nodeDataNext?.name || /* istanbul ignore next */ '',
            ...iconData,
            identifier: id,
            skipCondition: nodeDataNext?.skipInfo?.evaluatedCondition ? nodeDataNext.skipInfo.skipCondition : undefined,
            when: nodeDataNext?.nodeRunInfo,
            status: nodeDataNext?.status as ExecutionStatus,
            type: getExecutionPipelineNodeType(nodeDataNext?.stepType),
            data: nodeDataNext
          },
          children: []
        } as PipelineGraphState)
      }
      const nextLevels = nodeAdjacencyListMap?.[id].nextIds
      if (nextLevels) {
        items.push(...processNodeDataV1(nextLevels, nodeMap, nodeAdjacencyListMap, rootNodes))
      }
    })
  })
  return items
}
