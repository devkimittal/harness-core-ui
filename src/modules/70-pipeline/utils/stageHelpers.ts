/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { defaultTo, get, isEmpty } from 'lodash-es'
import { getMultiTypeFromValue, MultiTypeInputType } from '@wings-software/uicore'
import type { GraphLayoutNode, PipelineExecutionSummary } from 'services/pipeline-ng'
import type { StringKeys } from 'framework/strings'
import type { GetExecutionStrategyYamlQueryParams, PipelineInfoConfig, StageElementConfig } from 'services/cd-ng'
import { ManifestDataType } from '@pipeline/components/ManifestSelection/Manifesthelper'
import type { ManifestTypes } from '@pipeline/components/ManifestSelection/ManifestInterface'
import type { InputSetDTO } from './types'
import type { DeploymentStageElementConfig, PipelineStageWrapper, StageElementWrapper } from './pipelineTypes'

export enum StageType {
  DEPLOY = 'Deployment',
  BUILD = 'CI',
  FEATURE = 'FeatureFlag',
  PIPELINE = 'Pipeline',
  APPROVAL = 'Approval',
  CUSTOM = 'Custom',
  Template = 'Template',
  SECURITY = 'Security'
}

export enum ServiceDeploymentType {
  Kubernetes = 'Kubernetes',
  NativeHelm = 'NativeHelm',
  amazonEcs = 'amazonEcs',
  amazonAmi = 'amazonAmi',
  awsCodeDeploy = 'awsCodeDeploy',
  winrm = 'winrm',
  awsLambda = 'awsLambda',
  pcf = 'pcf',
  ssh = 'ssh',
  ServerlessAwsLambda = 'ServerlessAwsLambda',
  ServerlessAzureFunction = 'ServerlessAzureFunctions',
  ServerlessGoogleFunctions = 'ServerlessGoogleFunctions',
  AmazonSAM = 'AwsSAM',
  AzureFunctions = 'AzureFunctions'
}

export const changeEmptyValuesToRunTimeInput = (inputset: any, propertyKey: string): InputSetDTO => {
  if (inputset) {
    Object.keys(inputset).forEach(key => {
      if (typeof inputset[key] === 'object') {
        changeEmptyValuesToRunTimeInput(inputset[key], key)
      } else if (inputset[key] === '' && ['tags'].indexOf(propertyKey) === -1) {
        inputset[key] = '<+input>'
      }
    })
  }
  return inputset
}

export function isCDStage(node?: GraphLayoutNode): boolean {
  return node?.nodeType === StageType.DEPLOY || node?.module === 'cd' || !isEmpty(node?.moduleInfo?.cd)
}

export function isCIStage(node?: GraphLayoutNode): boolean {
  return node?.nodeType === StageType.BUILD || node?.module === 'ci' || !isEmpty(node?.moduleInfo?.ci)
}

export function hasCDStage(pipelineExecution?: PipelineExecutionSummary): boolean {
  return pipelineExecution?.modules?.includes('cd') || !isEmpty(pipelineExecution?.moduleInfo?.cd)
}

export function hasCIStage(pipelineExecution?: PipelineExecutionSummary): boolean {
  return pipelineExecution?.modules?.includes('ci') || !isEmpty(pipelineExecution?.moduleInfo?.ci)
}

export const getHelperTextString = (
  invalidFields: string[],
  getString: (key: StringKeys) => string,
  isServerlessDeploymentTypeSelected = false
): string => {
  return `${invalidFields.length > 1 ? invalidFields.join(', ') : invalidFields[0]} ${
    invalidFields.length > 1 ? ' are ' : ' is '
  } ${
    isServerlessDeploymentTypeSelected
      ? getString('pipeline.artifactPathDependencyRequired')
      : getString('pipeline.tagDependencyRequired')
  }`
}

export const getHelpeTextForTags = (
  fields: {
    imagePath?: string
    artifactPath?: string
    region?: string
    connectorRef: string
    registryHostname?: string
    repository?: string
    repositoryPort?: number
    artifactDirectory?: string
  },
  getString: (key: StringKeys) => string,
  isServerlessDeploymentTypeSelected = false
): string => {
  const {
    connectorRef,
    region,
    imagePath,
    artifactPath,
    registryHostname,
    repository,
    repositoryPort,
    artifactDirectory
  } = fields
  const invalidFields: string[] = []
  if (!connectorRef || getMultiTypeFromValue(connectorRef) === MultiTypeInputType.RUNTIME) {
    invalidFields.push(getString('connector'))
  }
  if (region !== undefined && (!region || getMultiTypeFromValue(region) === MultiTypeInputType.RUNTIME)) {
    invalidFields.push(getString('regionLabel'))
  }
  if (
    registryHostname !== undefined &&
    (!registryHostname || getMultiTypeFromValue(registryHostname) === MultiTypeInputType.RUNTIME)
  ) {
    invalidFields.push(getString('connectors.GCR.registryHostname'))
  }
  if (
    !isServerlessDeploymentTypeSelected &&
    (!imagePath || getMultiTypeFromValue(imagePath) === MultiTypeInputType.RUNTIME)
  ) {
    invalidFields.push(getString('pipeline.imagePathLabel'))
  }
  if (
    !isServerlessDeploymentTypeSelected &&
    (!imagePath || getMultiTypeFromValue(artifactPath) === MultiTypeInputType.RUNTIME)
  ) {
    invalidFields.push(getString('pipeline.artifactPathLabel'))
  }
  if (repository !== undefined && (!repository || getMultiTypeFromValue(repository) === MultiTypeInputType.RUNTIME)) {
    invalidFields.push(getString('repository'))
  }
  if (
    repositoryPort !== undefined &&
    (!repositoryPort || getMultiTypeFromValue(repositoryPort) === MultiTypeInputType.RUNTIME)
  ) {
    invalidFields.push(getString('pipeline.artifactsSelection.repositoryPort'))
  }
  if (
    isServerlessDeploymentTypeSelected &&
    (!artifactDirectory || getMultiTypeFromValue(artifactDirectory) === MultiTypeInputType.RUNTIME)
  ) {
    invalidFields.push(getString('pipeline.artifactsSelection.artifactDirectory'))
  }

  const helpText = getHelperTextString(invalidFields, getString, isServerlessDeploymentTypeSelected)

  return invalidFields.length > 0 ? helpText : ''
}

export const isServerlessDeploymentType = (deploymentType: string): boolean => {
  return (
    deploymentType === ServiceDeploymentType.ServerlessAwsLambda ||
    deploymentType === ServiceDeploymentType.ServerlessAzureFunction ||
    deploymentType === ServiceDeploymentType.ServerlessGoogleFunctions ||
    deploymentType === ServiceDeploymentType.AmazonSAM ||
    deploymentType === ServiceDeploymentType.AzureFunctions
  )
}

export const isServerlessManifestType = (selectedManifest: ManifestTypes | null): boolean => {
  return selectedManifest === ManifestDataType.ServerlessAwsLambda
}

export const getSelectedDeploymentType = (
  stage: StageElementWrapper<DeploymentStageElementConfig> | undefined,
  getStageFromPipeline: <T extends StageElementConfig = StageElementConfig>(
    stageId: string,
    pipeline?: PipelineInfoConfig | undefined
  ) => PipelineStageWrapper<T>,
  isPropagating = false
): GetExecutionStrategyYamlQueryParams['serviceDefinitionType'] => {
  if (isPropagating) {
    const parentStageId = get(stage, 'stage.spec.serviceConfig.useFromStage.stage', null)
    const parentStage = getStageFromPipeline<DeploymentStageElementConfig>(defaultTo(parentStageId, ''))
    return get(parentStage, 'stage.stage.spec.serviceConfig.serviceDefinition.type', null)
  }
  return get(stage, 'stage.spec.serviceConfig.serviceDefinition.type', null)
}
