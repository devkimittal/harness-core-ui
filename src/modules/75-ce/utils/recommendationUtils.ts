/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

export const addBufferToValue = (value: number, bufferPercentage: number): number =>
  +(((100 + bufferPercentage) / 100) * value).toFixed(2)

export const calculateSavingsPercentage = (savings: number, totalCost: number): string =>
  `(${Math.floor((savings / totalCost) * 100)}%)`

export const isResourceConsistent = (sumCpu: number, sumMemory: number, maxCpu: number, maxMemory: number): boolean => {
  const isInconsistent = Math.round(sumCpu) < Math.round(maxCpu) || Math.round(sumMemory) < Math.round(maxMemory)

  const anyZero =
    Math.round(sumCpu) == 0 || Math.round(sumMemory) == 0 || Math.round(maxCpu) == 0 || Math.round(maxMemory) == 0

  return !isInconsistent || !anyZero
}

export const calculateNodes = (
  sumCpu: number,
  sumMemory: number,
  maxCpu: number,
  maxMemory: number,
  minNodes: number
): { maximumNodes: number; minimumNodes: number } => {
  let minimumNodes = minNodes || 3

  let maximumNodes = Math.min(Math.floor(sumCpu / maxCpu), Math.floor(sumMemory / maxMemory))
  maximumNodes = Math.max(maximumNodes, 1)

  if (maximumNodes < minNodes) {
    minimumNodes = maximumNodes
  }

  return { maximumNodes, minimumNodes }
}
