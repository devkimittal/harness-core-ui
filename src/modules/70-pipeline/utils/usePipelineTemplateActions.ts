/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { defaultTo, isEqual } from 'lodash-es'
import { parse } from 'yaml'
import produce from 'immer'
import { useCallback } from 'react'
import type { PipelineInfoConfig } from 'services/cd-ng'
import type { TemplateSummaryResponse } from 'services/template-ng'
import { usePipelineContext } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import { useTemplateSelector } from '@pipeline/utils/useTemplateSelector'
import { createTemplate } from '@pipeline/utils/templateUtils'
import { getIdentifierFromValue } from '@common/components/EntityReference/EntityReference'

interface TemplateActionsReturnType {
  onUseTemplate: (template: TemplateSummaryResponse, isCopied?: boolean) => Promise<void>
  onRemoveTemplate: () => Promise<void>
  onOpenTemplateSelector: () => void
}

export function usePipelineTemplateActions(): TemplateActionsReturnType {
  const {
    state: { pipeline, templateTypes },
    setTemplateTypes,
    updatePipeline
  } = usePipelineContext()
  const { openTemplateSelector, closeTemplateSelector } = useTemplateSelector()

  const onUseTemplate = useCallback(
    async (template: TemplateSummaryResponse, isCopied = false) => {
      closeTemplateSelector()
      if (
        !isCopied &&
        isEqual(pipeline?.template?.templateRef, template.identifier) &&
        isEqual(pipeline?.template?.versionLabel, template.versionLabel)
      ) {
        return
      }
      const processNode = isCopied
        ? produce(defaultTo(parse(template?.yaml || '')?.template.spec, {}) as PipelineInfoConfig, draft => {
            draft.name = defaultTo(pipeline?.name, '')
            draft.identifier = defaultTo(pipeline?.identifier, '')
          })
        : createTemplate(pipeline, template)
      await updatePipeline(processNode)
      if (!isCopied && template?.identifier && template?.childType) {
        templateTypes[template.identifier] = template.childType
        setTemplateTypes(templateTypes)
      }
    },
    [closeTemplateSelector, pipeline, updatePipeline]
  )

  const onOpenTemplateSelector = useCallback(() => {
    openTemplateSelector({
      templateType: 'Pipeline',
      selectedTemplateRef: getIdentifierFromValue(defaultTo(pipeline.template?.templateRef, '')),
      selectedVersionLabel: pipeline.template?.versionLabel,
      onUseTemplate
    })
  }, [pipeline.template, openTemplateSelector, onUseTemplate])

  const onRemoveTemplate = useCallback(async () => {
    const node = pipeline
    const processNode = produce({} as PipelineInfoConfig, draft => {
      draft.name = defaultTo(node?.name, '')
      draft.identifier = defaultTo(node?.identifier, '')
    })
    await updatePipeline(processNode)
  }, [pipeline, updatePipeline])

  return { onUseTemplate, onRemoveTemplate, onOpenTemplateSelector }
}
