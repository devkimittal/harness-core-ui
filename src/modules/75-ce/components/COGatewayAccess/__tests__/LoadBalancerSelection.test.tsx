/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import type { AccessPoint, AccessPointCore } from 'services/lw'
import * as lwServices from 'services/lw'
import LoadBalancerSelection from '../LoadBalancerSelection'
import { initialGatewayDetails } from './data'

const mockAccessPointList = {
  response: [
    {
      name: 'mock.com',
      id: 'mock.com',
      host_name: 'mock.com',
      metadata: {
        albArn: 'mockalbARN'
      }
    } as AccessPoint
  ]
}

const mockAccessPointResourceData = {
  response: [
    {
      details: {
        albARN: 'mockalbARN',
        name: 'mockALBname',
        id: 'mock.com'
      }
    }
  ] as AccessPointCore[]
}

jest.mock('services/lw', () => ({
  useListAccessPoints: jest.fn().mockImplementation(() => ({
    data: mockAccessPointList,
    loading: false,
    refetch: jest.fn(() => mockAccessPointList)
  })),
  useAccessPointResources: jest.fn().mockImplementation(() => ({
    data: mockAccessPointResourceData,
    loading: false,
    refetch: jest.fn(() => mockAccessPointResourceData)
  }))
}))

describe('load balancer selection cases', () => {
  test('no data available for access points and supported resources', () => {
    jest.spyOn(lwServices, 'useListAccessPoints').mockImplementation(
      () =>
        ({
          data: null,
          loading: false,
          refetch: jest.fn()
        } as any)
    )
    jest.spyOn(lwServices, 'useAccessPointResources').mockImplementation(
      () =>
        ({
          data: null,
          loading: false,
          refetch: jest.fn()
        } as any)
    )

    const { container } = render(
      <TestWrapper>
        <LoadBalancerSelection gatewayDetails={initialGatewayDetails} setGatewayDetails={jest.fn()} />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
  })
})
