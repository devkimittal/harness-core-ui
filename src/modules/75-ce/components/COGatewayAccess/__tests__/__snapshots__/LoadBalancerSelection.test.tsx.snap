// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load balancer selection cases no data available for access points and supported resources 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main dnsLinkContainer"
  >
    <h3
      class="StyledProps--font StyledProps--font-h3 StyledProps--main header StyledProps--font-weight-bold"
      level="3"
    >
      ce.co.autoStoppingRule.setupAccess.selectLb
    </h3>
    <div
      class="Layout--horizontal StyledProps--main selectLoadBalancerContainer"
    >
      <div>
        <p
          class="StyledProps--font StyledProps--main helpText"
        >
          ce.co.autoStoppingRule.setupAccess.selectLbHelpText
        </p>
        <div
          class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-alignItems-center StyledProps--flex-justifyContent-flex-start"
        >
          <div
            class="bp3-popover-wrapper loadBalancerSelector Select--main bp3-fill"
          >
            <div
              class="bp3-popover-target"
            >
              <div
                class="bp3-input-group"
              >
                <input
                  autocomplete="off"
                  class="bp3-input"
                  name="accessPoint"
                  placeholder="- Select -"
                  style="padding-right: 0px;"
                  type="text"
                  value=""
                />
                <span
                  class="bp3-input-action"
                >
                  <span
                    class="bp3-icon bp3-icon-chevron-down StyledProps--main StyledProps--padding-small"
                    icon="chevron-down"
                  >
                    <svg
                      data-icon="chevron-down"
                      height="14"
                      viewBox="0 0 16 16"
                      width="14"
                    >
                      <desc>
                        chevron-down
                      </desc>
                      <path
                        d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
          <p
            class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-primary6"
            style="cursor: pointer;"
          >
            +ce.co.accessPoint.new
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
